parley/src/layout/data.rs

```diff
@@ -287,176 +352,120 @@ impl<B: Brush> LayoutData<B> {
-        }
-        let coords_end = self.coords.len();
+
         let mut run = RunData {
             font_index,
             font_size,
             synthesis,
             coords_range: coords_start..coords_end,
-            text_range: 0..0,
+            text_range,
             bidi_level,
-            ends_with_newline: false,
             cluster_range,
             glyph_start: self.glyphs.len(),
-            metrics: RunMetrics {
-                ascent: metrics.ascent,
-                descent: metrics.descent,
-                leading: metrics.leading,
-                underline_offset: metrics.underline_offset,
-                underline_size: metrics.stroke_size,
-                strikethrough_offset: metrics.strikeout_offset,
-                strikethrough_size: metrics.stroke_size,
-            },
+            metrics,
             word_spacing,
             letter_spacing,
             advance: 0.,
         };
-        // Track these so that we can flush if they overflow a u16.
-        let mut glyph_count = 0_usize;
-        let mut text_offset = 0;
-        macro_rules! flush_run {
-            () => {
-                if !run.cluster_range.is_empty() {
-                    self.runs.push(run.clone());
-                    self.items.push(LayoutItem {
-                        kind: LayoutItemKind::TextRun,
-                        index: self.runs.len() - 1,
-                        bidi_level: run.bidi_level,
-                    });
-                    run.text_range = text_offset..text_offset;
-                    run.cluster_range.start = run.cluster_range.end;
-                    run.glyph_start = self.glyphs.len();
-                    run.advance = 0.;
-                    glyph_count = 0;
-                }
-            };
+
+        // HarfBuzz returns glyphs in visual order, so we need to process them as such while
+        // maintaining logical ordering of clusters.
+
+        let glyph_infos = glyph_buffer.glyph_infos();
+        if glyph_infos.is_empty() {
+            return;
+        }
+        let glyph_positions = glyph_buffer.glyph_positions();
+        let scale_factor = font_size / units_per_em;
+        let cluster_range_start = self.clusters.len();
+        let is_rtl = bidi_level & 1 == 1;
+        if !is_rtl {
+            run.advance = process_clusters(
+                &mut self.clusters,
+                &mut self.glyphs,
+                scale_factor,
+                glyph_infos,
+                glyph_positions,
+                char_infos,
+                source_text.char_indices(),
+            );
+        } else {
+            let mut clusters = core::mem::take(&mut self.scratch_clusters);
+            run.advance = process_clusters(
+                &mut clusters,
+                &mut self.glyphs,
+                scale_factor,
+                glyph_infos,
+                glyph_positions,
+                char_infos,
+                source_text.char_indices().rev(),
+            );
+
```

Comment:
Would https://doc.rust-lang.org/stable/std/primitive.slice.html#method.reverse to reverse a mutable slice in-place work here?